### PR TITLE
Allow code coverage opt-in for live tests

### PR DESCRIPTION
**Pull request by benbp**
This PR re-works code coverage inclusion to be a little more flexible (instead of cherry-picking a test matrix scenario in the yaml, mark a coverage variable in the matrix itself). Additionally, it allows opt-in of coverage results in the live tests where desired (for example, the azure communication services team wants live test coverage for select clouds).

### CODE REVIEW

File: eng/pipelines/templates/steps/code-coverage.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      jdkArchitectureOption: 'x64'
+      goals: jacoco:report-aggregate
+
+  # Azure DevOps only seems to respect the last code coverage result published, so only do this for Windows + Java LTS.
+  # Code coverage reporting is setup only for Track 2 modules.
+  - task: PublishCodeCoverageResults@1
+    condition: eq(variables['Coverage'], 'true')
+    inputs:
+      codeCoverageTool: JaCoCo
+      summaryFileLocation: sdk/${{ parameters.ServiceDirectory }}/target/site/jacoco-aggregate/jacoco.xml
```
Comment (JianpingChen) [inline eng/pipelines/templates/steps/code-coverage.yml:20]: This will work for jacoco:report-aggregate but may not work for communication since each sdk in communication has its own pipeline and it runs jacoco:report instead of jacoco:report-aggregate.  The summaryFileLocation will be inside the sdk instead of the serviceDirectory

File: eng/pipelines/templates/steps/code-coverage.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    displayName: 'Generate aggregate code coverage report'
+    condition: eq(variables['Coverage'], 'true')
+    inputs:
+      mavenPomFile: sdk/${{ parameters.ServiceDirectory }}/pom.xml
+      options: -Pcoverage
+      mavenOptions: '$(MemoryOptions) $(LoggingOptions)'
+      javaHomeOption: 'JDKVersion'
+      jdkVersionOption: $(JavaTestVersion)
+      jdkArchitectureOption: 'x64'
+      goals: jacoco:report-aggregate
```
Comment (JianpingChen) [inline eng/pipelines/templates/steps/code-coverage.yml:12]: I'm not sure about this one. What happens if the corresponding pom.xml doesn't have "modules" section? Will jacoco:report-aggregate still work?

File: eng/pipelines/templates/jobs/live.tests.yml [hunk 3]
```diff
@@ -144,3 +144,5 @@ jobs:
           testRunTitle: 'Live tests for ${{ parameters.ServiceDirectory }} ${{ parameters.CloudConfig.Cloud }} $(Agent.JobName)'
           ${{ if ne(parameters.TestResultsFiles, '') }}:
             testResultsFiles: ${{ parameters.TestResultsFiles }}
+
+      - template: ../steps/code-coverage.yml
```
Comment (JianpingChen) [inline eng/pipelines/templates/jobs/live.tests.yml:148]: Communication will still fail in INT since in the pom.xml for communication sdks, jacoco is plugged into "verify" which is triggered in the task above, "'Run tests: $(Agent.JobName)'"

File: eng/pipelines/templates/steps/code-coverage.yml (hunk 2 above)
Comment (JianpingChen) [reply eng/pipelines/templates/steps/code-coverage.yml:12]: Plus, Jacoco is plugged into "verify" already, there is no need to run this separately if "verify" was run already

**Review by JianpingChen — CHANGES_REQUESTED** [review]

File: eng/pipelines/templates/stages/platform-matrix.json [hunk 4]
```diff
@@ -1,6 +1,7 @@
 {
   "displayNames": {
-    "true": "TestFromSource"
+      "false": "",
```
Comment (weshaggard) [inline eng/pipelines/templates/stages/platform-matrix.json:3]: What cases would we need to nullify false in the display name?

Comment (weshaggard) [reply eng/pipelines/templates/stages/platform-matrix.json:3]: I think I understand that you are doing here now. Why not use a similar pattern to TestFromSourceConfig and next the coverage option? While this mapping works right now if we ever have any other true/false values this might map them as well.

Comment (benbp) [reply eng/pipelines/templates/stages/platform-matrix.json:3]: The problem is this value is made to be overridden/updated by different test scenarios. And due to the "grouping name" problem, it will list the job display name as "Coverage" regardless of whether the value is false or true.

Looks like this PR may not be the best attempt regardless, so a moot point here, but this is an annoying side effect of auto-name generation+using a simple config with less metadata support per field. I still don't have a great solution.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml [hunk 5]
```diff
@@ -68,10 +68,8 @@ stages:
           - ${{ if eq(parameters.SDKType, 'data') }}:
             - TestFromSource=^(?!true).*
         MatrixReplace:
-          # Temporarily replace macOS agents with ubuntu agents because of ongoing pool capacity issues
```
Comment (mitchdenny) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:71]: This should probably be a separate PR.

File: eng/pipelines/templates/steps/code-coverage.yml (hunk 2 above)
Comment (mitchdenny) [reply eng/pipelines/templates/steps/code-coverage.yml:12]: I don't think this is a change to existing behavior, just the template pieces being moved.

File: eng/pipelines/templates/stages/platform-matrix.json (hunk 4 above)
Comment (weshaggard) [reply eng/pipelines/templates/stages/platform-matrix.json:3]: I wonder if we can come up with some sort of display name format string. Not sure if it will work or handle all the cases, or really deal with conflicts but it is just me brainstorming.

Comment (benbp) [reply eng/pipelines/templates/stages/platform-matrix.json:3]: Hmm, riffing on that idea, I wonder if adding support for JSON comments would give us a a convenient place to do code-gen hinting and formatting without over-complicating the config schema, e.g.

```
// use key
{
  "matrix": {
    "Coverage": "true"  //+matrix:displayName:$key
  }
}

// use literal
{
  "matrix": {
    "Coverage": "true"  //+matrix:displayName:EnableCoverage
  }
}

// use formatted string with key and value
{
  "matrix": {
    "Coverage": "true"  //+matrix:displayName:$key_$value
  }
}
```

Comment (benbp) [reply eng/pipelines/templates/stages/platform-matrix.json:3]: An alternative more along what you suggested:

```
// Replace display name for Coverage variable with dynamic format string
{
  "displayNames": {
    "Coverage": "$key_$value"
  },
  "matrix": {
    "Coverage": "true"
  }
}

```

Though I've always felt like `displayNames` as a top-level key was somewhat out of place and ruined the aspect of the config where every key had a direct relationship to the actual matrix, as opposed to an auxiliary purpose.

Comment (weshaggard) [reply eng/pipelines/templates/stages/platform-matrix.json:3]: We can keep thinking about the display name outside the scope of this PR. I can live with what you have here for this case.

**Review by weshaggard — APPROVED** [review]
Looks reasonable
